File: PhotoViewer-ObjC/DataSource.m
//
//  DataSource.m
//  PhotoViewer
//
//  Created by ASIM27 on 1/20/17.
//  Copyright © 2017 km3h. All rights reserved.
//

#import "DataSource.h"
#import "SessionDelegate.h"

@interface DataSource ()
@property (strong, nonatomic) NSURLRequest *urlRequest;
@property (strong, nonatomic) NSURLRequest *urlRequestOffline;
@property (weak, nonatomic) NSURLSession *urlSession;
@property (nonatomic, copy, nullable) void (^photoDictionaryCompletionHandler) (NSDictionary *);
@property (nonatomic, copy, nullable) void (^taskCompletionHandler) (NSData*, NSURLResponse*, NSError*);
@property (nonatomic, copy, nullable) void (^taskOfflineCompletionHandler) (NSData*, NSURLResponse*, NSError*);

@end

@implementation DataSource

-(instancetype)init
{
    self = [super init];
    [self createPhotoDirectory];
    
    __weak DataSource *weakSelf = self;
    self.taskCompletionHandler = ^(NSData *data, NSURLResponse *response, NSError *error) {
        [weakSelf taskProcessor:data response:response error:error];
    };
    
    self.taskOfflineCompletionHandler = ^(NSData *data, NSURLResponse *response, NSError *error) {
        [weakSelf taskProcessorOffline:data response:response error:error];
    };
    
    return self;
}

-(void)setUrlString:(NSString *)urlString
{
    self->_urlString = urlString;
    NSURL *url = [NSURL URLWithString:urlString];
    self.urlRequest = [NSURLRequest requestWithURL:url];
    self.urlRequestOffline = [NSURLRequest requestWithURL:url cachePolicy:NSURLRequestReturnCacheDataElseLoad timeoutInterval:10.0];
}

-(void)createPhotoDirectory
{
    NSURL *documentsURL = [[[NSFileManager defaultManager] URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask] firstObject];
    NSString *dataPath = [documentsURL URLByAppendingPathComponent:@"images"].path;
    
    if ([[NSFileManager defaultManager] fileExistsAtPath:dataPath])
    {
        return;
    }
    else
    {
        [[NSFileManager defaultManager] createDirectoryAtPath:dataPath withIntermediateDirectories:NO attributes:nil error:NULL];
    }
}

-(void)genericTaskProcessor:(NSData*)data response:(NSURLResponse*)urlResponse error:(NSError*)error
{
    if ([urlResponse isKindOfClass:[NSHTTPURLResponse class]])
    {
        NSInteger statusCode = ((NSHTTPURLResponse*)urlResponse).statusCode;
        if (statusCode == 200 || statusCode == 304)
        {
            if (data == nil)
            {
                return;
            }
            NSArray *json = [self serializeJson:data];
            self.photoDictionaryCompletionHandler([self transformData:json]);
        }
    }
}

-(void)taskProcessor:(NSData*)data response:(NSURLResponse*)urlResponse error:(NSError*)error
{
    if (error)
    {
        NSLog(@"error %@", error);
    }
    
    if (urlResponse != nil )
    {
        [self genericTaskProcessor:data response: urlResponse error: error];
    }
    else
    {
        NSURLSessionDataTask *dataTask;
        switch (self.urlSessionType)
        {
            case URLSessionTypeDefault:
            case URLSessionTypeDelegate:
                dataTask = [[self getSessionWithCompletionHandler:self.taskOfflineCompletionHandler] dataTaskWithRequest:self.urlRequestOffline];
                break;
            case URLSessionTypeCompletionHandler:
                dataTask = [[NSURLSession sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]]  dataTaskWithRequest:self.urlRequestOffline completionHandler:self.taskOfflineCompletionHandler];
                break;
            default:
                break;
        }
        [dataTask resume];
    }
}

-(void)taskProcessorOffline:(NSData*)data response:(NSURLResponse*)urlResponse error:(NSError*)error
{
    if (error)
    {
        NSLog(@"error %@", error);
    }
    
    if (urlResponse != nil )
    {
        [self genericTaskProcessor:data response: urlResponse error: error];
    }
}

-(void)photoDictionary:(void (^) (NSDictionary *)) completionHandler
{
    NSURLSessionDataTask *dataTask;
    self.photoDictionaryCompletionHandler = completionHandler;
    
    switch (self.urlSessionType)
    {
        case URLSessionTypeDefault:
        case URLSessionTypeDelegate:
            dataTask = [[self getSessionWithCompletionHandler:self.taskCompletionHandler] dataTaskWithRequest:self.urlRequest];
            break;
        case URLSessionTypeCompletionHandler:
            dataTask = [[NSURLSession sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]]  dataTaskWithRequest:self.urlRequest completionHandler:self.taskCompletionHandler];
            break;
        default:
            break;
    }
    
    [dataTask resume];
}
-(NSURLSession*)getSessionWithCompletionHandler:(void (^) (NSData*, NSURLResponse*, NSError*))completionHandler
{
    SessionDelegate *delegate = [[SessionDelegate alloc] initWithCompletionHandler:completionHandler];
    NSURLSession *session = [NSURLSession sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration] delegate:delegate delegateQueue:nil];
    return session;
}

-(NSArray*)serializeJson:(NSData*)data
{
    if (data != nil)
    {
        NSArray *json = (NSArray*)[NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingAllowFragments error:NULL];
        return json;
    }
    return nil;
}

-(NSDictionary*)transformData:(NSArray*)json
{
    NSMutableDictionary *transformedData = [NSMutableDictionary dictionaryWithCapacity:json.count];
    
    for (NSDictionary *photo in json)
    {
        NSNumber *albumName = photo[@"albumId"];
        NSMutableArray *photoArray = transformedData[albumName];
        
        if (photoArray == nil)
        {
            photoArray = [NSMutableArray array];
            transformedData[albumName] = photoArray;
        }
        [photoArray addObject:photo];
    }
    return transformedData;
}

-(void)photo:(DataSourcePhotoType)photoType photo:(NSDictionary*)photo completionHandler: (void (^) (UIImage*)) completionHandler
{
    NSString *photoName = photo[@"id"];
    NSURL *documentsURL = [[[NSFileManager defaultManager] URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask] firstObject];
    
    NSString *filePath;
    NSString *photoPath;
    NSString *photoSize;
    
    if  (photoType == DataSourcePhotoTypeThumnail)
    {
        photoSize = @"thumbnail";
        photoPath = photo[@"thumbnailUrl"];
    }
    else
    {
        photoSize = @"fullPhoto";
        photoPath = photo[@"url"];
    }
    
    NSString *pathComponent = [NSString stringWithFormat:@"images/%@_%@.png", photoName, photoSize];
    filePath = [documentsURL URLByAppendingPathComponent:pathComponent].path;
    
    if ([[NSFileManager defaultManager] fileExistsAtPath:filePath])
    {
        completionHandler([UIImage imageWithContentsOfFile:filePath]);
    }
    else
    {
        NSURL *fileURL = [NSURL fileURLWithPath:filePath];
        dispatch_async(dispatch_get_global_queue( DISPATCH_QUEUE_PRIORITY_DEFAULT, 0),  ^{
            NSData *data = [NSData dataWithContentsOfURL:[NSURL URLWithString:photoPath]];
            UIImage *getImage = [UIImage imageWithData:data];
            [data writeToURL:fileURL atomically:YES];
            dispatch_async(dispatch_get_main_queue(), ^{
                completionHandler(getImage);
            });
        });
    }
}

@end

File: PhotoViewer/DataSource.swift
//
//  DataSource.swift
//  PhotoViewer
//
//  Created by ASIM27 on 11/16/16.
//  Copyright © 2016 km3h. All rights reserved.
//

import UIKit

enum DataSourcePhotoType
{
    case THUMBNAIL
    case FULLPHOTO
}

enum URLSessionType
{
    case Delegate
    case CompletionHandler
    case Default
}

class DataSource: NSObject
{
    var urlRequest : URLRequest?
    var urlRequestOffline : URLRequest?
    var shouldReadFromCache : Bool?
    public var urlSessionType : URLSessionType?
    
    typealias handlerName = ([NSNumber : [[String : AnyObject]]]?) -> Void
    var handler : handlerName?
    
    override init()
    {
        super.init()
        self.createPhotoDirectory()
    }
    
    public var urlString : String? {
        didSet {
            guard let urlString = self.urlString else {
                return
            }
            
            guard let url = URL(string: urlString) else {
                return
            }
            self.urlRequest = URLRequest(url: url)
            self.urlRequestOffline = URLRequest(url: url as URL, cachePolicy: .returnCacheDataElseLoad, timeoutInterval: 10.0)
        }
    }
    
    func createPhotoDirectory() -> Void
    {
        let documentsURL = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!
        let dataPath = documentsURL.appendingPathComponent("images").path
        
        if FileManager.default.fileExists(atPath: dataPath)
        {
            return
        }
        
        do
        {
            try FileManager.default.createDirectory(atPath: dataPath, withIntermediateDirectories: false, attributes: nil)
        }
        catch let error as NSError
        {
            print(error.localizedDescription);
        }
    }
    
    func genericTaskHandler(data :Data?, response : URLResponse?, error: Error?) -> Void
    {
        if let httpResponse = response as? HTTPURLResponse
        {
            let statusCode = httpResponse.statusCode
            if (statusCode == 200 || statusCode == 304)
            {
                do
                {
                    guard let data = data else
                    {
                        print("Error with Json1: \(error)")
                        return
                    }
                    
                    guard let json = try JSONSerialization.jsonObject(with: data, options:.allowFragments) as? [NSDictionary] else
                    {
                        print("Error with Json2: \(error)")
                        return
                    }
                    
                    guard let handler = self.handler else
                    {
                        print("Error with Json3: \(error)")
                        return
                    }
                    
                    handler(self.transformData(json: json))
                    
                }
                catch
                {
                    print("Error with Json4: \(error)")
                }
            }
        }
    }
    
    func taskHandler(data :Data?, response : URLResponse?, error: Error?) -> Void
    {
        if let error = error
        {
            print("error \(error)")
        }
        
        if (response as? HTTPURLResponse) != nil
        {
            self.genericTaskHandler(data: data, response: response, error: error)
        }
        else
        {
            let taskOffline = self.getSession(handler : taskHandlerOffline)?.dataTask(with: self.urlRequestOffline!)
            taskOffline?.resume() // read from nsurlcache if offline
        }
    }
    
    func taskHandlerOffline(data :Data?, response : URLResponse?, error: Error?) -> Void
    {
        if let error = error
        {
            print("error \(error)")
        }
        
        if (response as? HTTPURLResponse) != nil
        {
            self.genericTaskHandler(data: data, response: response, error: error)
        }
    }
    
    public func data(handler : @escaping ([NSNumber : [[String : AnyObject]]]?) -> Void)
    {
        var dataTask : URLSessionDataTask?
        self.handler = handler
        
        switch self.urlSessionType
        {
        case .some(.Delegate), .some(.Default):
            dataTask = self.getSession(handler : taskHandler)?.dataTask(with: self.urlRequest!)
        case .some(.CompletionHandler):
            dataTask = self.getSession(handler: taskHandler)?.dataTask(with: self.urlRequest!, completionHandler: taskHandler)
        default:
            print("default")
        }
        
        dataTask?.resume()
    }
    
    func getSession(handler : @escaping (Data?, URLResponse?, Error?) -> Void) -> URLSession?
    {
        return URLSession(configuration: .default, delegate: SessionDelegate(handler: handler), delegateQueue: nil)
    }
    
    func serializeJson(json data : Data?) -> [NSDictionary]?
    {
        if let data = data {
            do
            {
                let json = try JSONSerialization.jsonObject(with: data, options:.allowFragments) as? [NSDictionary]
                return json
            }
            catch
            {
                print("caught: \(error)")
                return nil
            }
        }
        return nil
    }
    
    func transformData(json : [NSDictionary]) -> [NSNumber : [[String : AnyObject]]]
    {
        var transformedData = [NSNumber : [[String : AnyObject]]]()
        
        for photo in json
        {
            if let albumName = photo["albumId"] as? NSNumber
            {
                if transformedData[albumName] != nil
                {
                    transformedData[albumName]!.append(photo as! [String : AnyObject])
                }
                else
                {
                    var photoArray = [[String : AnyObject]]()
                    photoArray.append(photo as! [String : AnyObject])
                    transformedData[albumName] = photoArray
                }
            }
        }
        return transformedData
    }
    
    public func photo (photoType: DataSourcePhotoType , photo : [String:AnyObject]?, handler : @escaping (UIImage?) -> Void)
    {
        guard let photoName = photo?["id"] else {
            return
        }
        
        let documentsURL = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!
        let filePath : String
        let photoPath : String
        if  photoType == .THUMBNAIL {
            filePath = documentsURL.appendingPathComponent("images/\(photoName)_thumbnail.png").path
            photoPath = photo!["thumbnailUrl"] as! String
        }
        else
        {
            filePath = documentsURL.appendingPathComponent("images/\(photoName)_fullPhoto.png").path
            photoPath = photo!["url"] as! String
        }
        
        if FileManager.default.fileExists(atPath: filePath)
        {
            handler(UIImage(contentsOfFile: filePath))
        }
        else
        {
            let fileURL = URL(fileURLWithPath: filePath)
            
            DispatchQueue.global(qos: DispatchQoS.background.qosClass).async {
                do
                {
                    let data = try Data(contentsOf: URL(string: photoPath)!)
                    let getImage = UIImage(data: data)
                    try data.write(to: fileURL)
                    DispatchQueue.main.async {
                        handler(getImage)
                    }
                }
                catch
                {
                    return
                }
            }
        }
    }
    
}
